File: 1367.linked-list-in-binary-tree.py
#
# @lc app=leetcode id=1367 lang=python3
#
# [1367] Linked List in Binary Tree
#

# @lc code=start
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def isSubPath(self, head: Optional[ListNode], root: Optional[TreeNode]) -> bool:
        def dfs(head: Optional[ListNode], node: Optional[TreeNode]) -> bool:
            if not head:
                return True  # Finished traversing the entire linked list
            if not node:
                return False  # Reached a leaf node without matching entire list
            if head.val == node.val:
                # Check either left or right subtree can continue matching
                return dfs(head.next, node.left) or dfs(head.next, node.right)
            return False  # Current values don't match; can't continue this path
        if not root:
            return False  # If root is null, no path can exist
        # Start DFS from current node or try from children (left/right) of current root
        return dfs(head, root) or self.isSubPath(head, root.left) or self.isSubPath(head, root.right)
# @lc code=end

File: 1367.linked-list-in-binary-tree.cpp
#
# @lc app=leetcode id=1367 lang=cpp
#
# [1367] Linked List in Binary Tree
#
# @lc code=start
/**
* Definition for singly-linked list.
* struct ListNode {
*     int val;
*     ListNode *next;
*     ListNode() : val(0), next(nullptr) {}
*     ListNode(int x) : val(x), next(nullptr) {}
*     ListNode(int x, ListNode *next) : val(x), next(next) {}
* };
*/
/**
* Definition for a binary tree node.
* struct TreeNode {
*     int val;
*     TreeNode *left;
*     TreeNode *right;
*     TreeNode() : val(0), left(nullptr), right(nullptr) {}
*     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
*     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
* };
*/
class Solution {
public:
    bool isSubPath(ListNode* head, TreeNode* root) {
        if (!root) return false;
        if (match(head, root)) return true;
        return isSubPath(head, root->left) || isSubPath(head, root->right);
    }
    
    bool match(ListNode* head, TreeNode* node) {
        if (!head) return true;
        if (!node) return false;
        if (head->val != node->val) return false;
        return match(head->next, node->left) || match(head->next, node->right);
    }
};
# @lc code=end